pkg/limits/partition_lifecycler.go

```diff
@@ -0,0 +1,118 @@
+package limits
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/twmb/franz-go/pkg/kgo"
+
+	kafka_partition "github.com/grafana/loki/v3/pkg/kafka/partition"
+)
+
+// PartitionLifecycler manages assignment and revocation of partitions.
+type PartitionLifecycler struct {
+	cfg              Config
+	partitionManager *PartitionManager
+	offsetManager    kafka_partition.OffsetManager
+	usage            *UsageStore
+	logger           log.Logger
+}
+
+// NewPartitionLifecycler returns a new PartitionLifecycler.
+func NewPartitionLifecycler(
+	cfg Config,
+	partitionManager *PartitionManager,
+	offsetManager kafka_partition.OffsetManager,
+	usage *UsageStore,
+	logger log.Logger,
+) *PartitionLifecycler {
+	return &PartitionLifecycler{
+		cfg:              cfg,
+		partitionManager: partitionManager,
+		offsetManager:    offsetManager,
+		usage:            usage,
+		logger:           logger,
+	}
+}
+
+// Assign implements kgo.OnPartitionsAssigned.
+func (l *PartitionLifecycler) Assign(ctx context.Context, _ *kgo.Client, topics map[string][]int32) {
+	for _, partitions := range topics {
+		l.partitionManager.Assign(ctx, partitions)
+		for _, partition := range partitions {
+			if err := l.checkOffsets(ctx, partition); err != nil {
+				level.Error(l.logger).Log(
+					"msg", "failed to check offsets, partition is ready",
+					"partition", partition,
+					"err", err,
+				)
+				l.partitionManager.SetReady(partition)
+			}
+		}
+		return
+	}
+}
+
+// Revoke implements kgo.OnPartitionsRevoked.
+func (l *PartitionLifecycler) Revoke(ctx context.Context, _ *kgo.Client, topics map[string][]int32) {
+	for _, partitions := range topics {
+		l.partitionManager.Revoke(ctx, partitions)
+		l.usage.EvictPartitions(partitions)
+		return
+	}
+}
+
+func (l *PartitionLifecycler) checkOffsets(ctx context.Context, partition int32) error {
```

Comment:
nit. This method does more than checking, maybe we come up with a name that implies setting the state.